crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs

```diff
@@ -0,0 +1,532 @@
+                        lines.push(DunderAllLine::JustAComment(LineWithJustAComment::new(
+                            comment, range,
+                        )));
+                    }
+                } else {
+                    lines.push(DunderAllLine::OneOrMoreItems(LineWithItems::new(
+                        &items_in_line,
+                        comment_in_line,
+                        range,
+                    )));
+                }
+                break;
+            }
+            Tok::NonLogicalNewline => {
+                if items_in_line.is_empty() {
+                    if let Some(comment) = comment_in_line {
+                        lines.push(DunderAllLine::JustAComment(LineWithJustAComment::new(
+                            comment, range,
+                        )));
+                        comment_in_line = None;
+                    }
+                } else {
+                    lines.push(DunderAllLine::OneOrMoreItems(LineWithItems::new(
+                        &items_in_line,
+                        comment_in_line,
+                        range,
+                    )));
+                    comment_in_line = None;
+                    items_in_line.clear();
+                }
+            }
+            Tok::Comment(_) => comment_in_line = Some(subrange),
+            Tok::String { value, .. } => items_in_line.push((value, subrange)),
+            Tok::Comma => continue,
+            _ => return None,
+        }
+    }
+    Some(lines)
+}
+
+#[derive(Debug)]
+struct LineWithJustAComment(TextRange);
+
+impl LineWithJustAComment {
+    fn new(comment_range: TextRange, total_dunder_all_range: TextRange) -> Self {
+        Self(comment_range + total_dunder_all_range.start())
+    }
+}
+
+#[derive(Debug)]
+struct LineWithItems {
+    items: Vec<(String, TextRange)>,
+    comment_range: Option<TextRange>,
+}
+
+impl LineWithItems {
+    fn new(
+        items: &[(String, TextRange)],
+        comment_range: Option<TextRange>,
+        total_dunder_all_range: TextRange,
+    ) -> Self {
+        assert!(
+            !items.is_empty(),
+            "Use the 'JustAComment' variant to represent lines with 0 items"
+        );
+        let offset = total_dunder_all_range.start();
+        Self {
+            items: items
+                .iter()
+                .map(|(s, r)| (s.to_owned(), r + offset))
+                .collect(),
+            comment_range: comment_range.map(|c| c + offset),
+        }
+    }
+}
+
+#[derive(Debug)]
+enum DunderAllLine {
+    JustAComment(LineWithJustAComment),
+    OneOrMoreItems(LineWithItems),
+}
+
+fn collect_dunder_all_items(lines: &[DunderAllLine]) -> Vec<DunderAllItem> {
+    // Given data on each line in `__all__`, group lines together into "items".
+    // Each item contains exactly one element,
+    // but might contain multiple comments attached to that element
+    // that must move with the element when `__all__` is sorted.
+    let mut all_items = vec![];
```

Comment:
I think a good approximation of the number of `items` is to use the number of lines if it's a multiline value and the number of items in the first line otherwise.. Initializing items with a specific capacity can help to reduce the number of times the vector needs to be resized.
```suggestion
    let mut all_items = Vec::with_capacity(match lines.as_slice() {
        [DunderAllLine::OneOrMoreItems(single)] => single.items.len(),
        _ => lines.len(),
    });
```